src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -118,15 +120,262 @@ void CoreWorkerDirectTaskSubmitter::AddWorkerLeaseClient(
+            // send the stolen tasks to that worker instead.
+            if (!thief_entry.PipelineToWorkerFull(max_tasks_in_flight_per_worker_)) {
+              // call OnWorkerIdle to ship the task to the thief
+              OnWorkerIdle(thief_addr, scheduling_key, /*error=*/false,
+                           thief_entry.assigned_resources);
+            } else {
+              // if the thief's pipeline is full, either find a new pseudo-thief, or send
+              // the stolen task to most recently used.
+
+              // If we have already found a pseudo thief, no need to look for another one
+              if (found_pseudo_thief &&
+                  worker_to_lease_entry_.find(rpc::WorkerAddress(pseudo_thief)) !=
+                      worker_to_lease_entry_.end()) {
+                auto &recipient_entry =
+                    worker_to_lease_entry_[rpc::WorkerAddress(pseudo_thief)];
+                if (!recipient_entry.PipelineToWorkerFull(
+                        max_tasks_in_flight_per_worker_)) {
+                  OnWorkerIdle(rpc::WorkerAddress(pseudo_thief), scheduling_key, false,
+                               recipient_entry.assigned_resources);
+                } else {
+                  // If the pseudo-thief's pipeline has become full, we will need to look
+                  // for a new one at the next iteration
+                  found_pseudo_thief = false;
+                }
+              } else {
+                // Find a worker with a number of tasks in flight that is less than the
+                // maximum value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle
+                // to send tasks to that worker
+                for (auto active_worker_addr : scheduling_key_entry.active_workers) {
+                  RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
+                            worker_to_lease_entry_.end());
+                  auto &recipient_entry = worker_to_lease_entry_[active_worker_addr];
+                  if (!recipient_entry.PipelineToWorkerFull(
+                          max_tasks_in_flight_per_worker_)) {
+                    found_pseudo_thief = true;
+                    pseudo_thief = active_worker_addr.ToProto();
+                    OnWorkerIdle(active_worker_addr, scheduling_key, false,
+                                 recipient_entry.assigned_resources);
+                    // If we find a worker with a non-full pipeline, all we need to do is
+                    // to submit the new task to the worker in question by calling
+                    // OnWorkerIdle once. We don't need to worry about other tasks in the
+                    // queue because the queue cannot have other tasks in it if there are
+                    // active workers with non-full pipelines.
+                    break;
+                  }
+                }
+              }
+            }
+          }
+          thief_entry.SetWorkerDoneStealing();
```

Comment:
It looks like this is duplicated.

Reply:
what do you mean by duplicated?

Reply:
Oh I meant that it looks like it gets called earlier in this function too.

Reply:
Yes, it's a bit confusing because of the many nested curly brackets, but the two SetWorkerDoneStealing() calls are one in the `if(...) {...}` and the `else {...}` branch respectively, so they are in sections of the code that never execute at the same time. The reason why we can't just call thief_entry.SetWorkerDoneStealing(); at the end of the if-statement is that the instruction needs to be called before we ReturnWorker